presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -284,6 +277,45 @@ protected Page computeNext()
         return types;
     }
 
+    /**
+     * Update memory usage with extra memory needed.
+     * The returned value indicates whether the reservation is within the limit
+     */
+    // TODO: update in the interface after the new memory tracking framework is landed
+    // Essentially we would love to have clean interfaces to support both pushing and pulling memory usage
+    // The following implementation is a hybrid model, where the push model is going to call the pull model causing reentrancy
+    private boolean updateMemoryWithYieldInfo()
+    {
+        long memorySize = getSizeInMemory();
+        if (partial) {
+            systemMemoryContext.setBytes(memorySize);
+            full = (memorySize > maxPartialMemory);
+        }
+        else {
+            // Operator/driver will be "blocked on memory" after we call setMemoryReservation.
+            // Once we return, this operator should be blocked until memory is available.
+            // After the operator gets to run again, we will call updateMemoryWithReservation again.
+            // It is fine since we are working with absolute numbers (not delta)．
+            if (!operatorContext.trySetMemoryReservation(memorySize)) {
+                // memorySize is the absolute memory (not delta) we need in order to proceed.
+                // If we do not have enough memory, we need to force a memory reservation and wait on it.
+                // Inform the caller that we cannot proceed to allocation more object if the memory is not available.
+                operatorContext.setMemoryReservation(memorySize);
+                // query context has GUARANTEED_MEMORY to make sure no query is blocked if it uses memory < 1MB.
+                // For such case, trySetMemoryReservation will fail but setMemoryReservation does not create a future to block the operator.
+                // So it is possible when we fail to reserve memory but still get unblocked.
+                return operatorContext.isWaitingForMemory().isDone();
+            }
+            // There is a chance that we are still waiting for memory and the above trySetMemoryReservation succeeds.
+            // This is because the memory we are trying to reserve is no greater than the memory we are waiting for.
+            // Need to inform the caller that we are still waiting for memory.
+            if (!operatorContext.isWaitingForMemory().isDone()) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
Revisit this in person．

I don't understand the comments. It also seems like the code itself can be simplified.